File: server/php/src/Infrastructure/Mappers/ExceptionMapper.php
<?php

namespace OnlinePayments\ExampleApp\Infrastructure\Mappers;

use OnlinePayments\ExampleApp\Application\Exceptions\SdkException;
use OnlinePayments\Sdk\ApiException;
use OnlinePayments\Sdk\DeclinedPaymentException;
use OnlinePayments\Sdk\ReferenceException;
use OnlinePayments\Sdk\ValidationException;
use OnlinePayments\Sdk\AuthorizationException;

class ExceptionMapper
{
    public static function map(\Exception $exception, ?string $customMessage = null): SdkException
    {
        if ($exception instanceof ValidationException) {
            return self::mapValidationException($exception);
        }

        if ($exception instanceof AuthorizationException) {
            return self::mapAuthorizationException($exception);
        }

        if ($exception instanceof ApiException
            || $exception instanceof DeclinedPaymentException
            || $exception instanceof ReferenceException) {
            return self::mapSdkErrorException($exception, $customMessage);
        }

        return new SdkException('An unexpected error occurred.', 500, $exception);
    }

    private static function mapValidationException(ValidationException $exception): SdkException
    {
        $message = self::extractValidationMessage($exception);
        return new SdkException($message, 400, $exception);
    }

    private static function mapAuthorizationException(AuthorizationException $exception): SdkException
    {
        return new SdkException('Invalid merchant data.', 403, $exception);
    }

    private static function mapSdkErrorException(\Exception $exception, ?string $customMessage = null): SdkException
    {
        $message = $customMessage ?? self::extractMessageFromSdkException($exception);
        $statusCode = self::extractStatusCodeFromSdkException($exception);

        return new SdkException($message, $statusCode, $exception);
    }

    private static function extractMessageFromSdkException(\Exception $exception): string
    {
        try {
            if (method_exists($exception, 'getErrors')) {
                $errors = $exception->getErrors();

                if (!empty($errors) && is_array($errors)) {
                    $firstError = $errors[0];

                    if (!empty($firstError->id)) {
                        return $firstError->id;
                    }

                    if (!empty($firstError->message)) {
                        return $firstError->message;
                    }

                    if (!empty($firstError->category) || !empty($firstError->errorCode)) {
                        return sprintf('%s (%s)', $firstError->category ?? '', $firstError->errorCode ?? '');
                    }
                }
            }

            return $exception->getMessage();
        } catch (\Exception $e) {
            return 'Error could not be retrieved.';
        }
    }

    private static function extractStatusCodeFromSdkException(\Exception $exception): int
    {
        try {
            if (method_exists($exception, 'getErrors')) {
                $errors = $exception->getErrors();

                if (!empty($errors) && is_array($errors)) {
                    $firstError = $errors[0];
                    $status = $firstError->httpStatusCode ?? 0;

                    if ($status > 0) {
                        return $status;
                    }
                }
            }

            return 422;
        } catch (\Exception $e) {
            return 422;
        }
    }

    private static function extractValidationMessage(ValidationException $exception): string
    {
        try {
            if (method_exists($exception, 'getErrors')) {
                $errors = $exception->getErrors();

                if (!empty($errors) && is_array($errors)) {
                    foreach ($errors as $error) {
                        if (!empty($error->id)) {
                            return $error->id;
                        }

                        if (!empty($error->message)) {
                            return $error->message;
                        }
                    }
                }
            }

            return $exception->getMessage();
        } catch (\Exception $e) {
            return 'Validation error occurred.';
        }
    }
}

File: server/python/app/infrastructure/mappers/exception_mapper.py
from onlinepayments.sdk.validation_exception import ValidationException as SdkValidationException
from onlinepayments.sdk.authorization_exception import AuthorizationException
from onlinepayments.sdk.api_exception import ApiException
from onlinepayments.sdk.declined_payment_exception import DeclinedPaymentException
from onlinepayments.sdk.reference_exception import ReferenceException
from app.application.exceptions.sdk_exception import SdkException

class ExceptionMapper:
    @staticmethod
    def map(exception: Exception, custom_message: str = None) -> SdkException:
        if isinstance(exception, SdkValidationException):
            return SdkException(ExceptionMapper._extract_message(exception), 400, exception)
        if isinstance(exception, AuthorizationException):
            return SdkException("Invalid merchant data.", 403, exception)
        if isinstance(exception, (ApiException, DeclinedPaymentException, ReferenceException)):
            return ExceptionMapper._map_sdk_error_exception(exception, custom_message)
        return SdkException("An unexpected error occurred.", 500, exception)

    @staticmethod
    def _map_sdk_error_exception(exception: Exception, custom_message: str = None) -> SdkException:
        message = custom_message or ExceptionMapper._extract_message(exception)
        status_code = ExceptionMapper._extract_status_code(exception)
        return SdkException(message, status_code, exception)

    @staticmethod
    def _extract_message(exception: Exception) -> str:
        try:
            if hasattr(exception, 'errors') and exception.errors:
                first = exception.errors[0]
                if getattr(first, 'id', None):
                    return first.id
                if getattr(first, 'message', None):
                    return first.message
                category = getattr(first, 'category', '') or ''
                error_code = getattr(first, 'error_code', '') or ''
                return f"{category} ({error_code})"
            return str(exception)
        except Exception:
            return "Error could not be retrieved."

    @staticmethod
    def _extract_status_code(exception: Exception) -> int:
        try:
            if hasattr(exception, 'status_code') and exception.status_code:
                return exception.status_code
            if hasattr(exception, 'errors') and exception.errors:
                first = exception.errors[0]
                status = getattr(first, 'http_status_code', 0) or 0
                if status > 0:
                    return status
            return 422
        except Exception:
            return 422